packages/api/src/did/did-doc.ts

```diff
@@ -0,0 +1,40 @@
+import { z } from 'zod'
+
+export const verificationMethod = z.object({
+  id: z.string(),
+  type: z.string(),
+  controller: z.string(),
+  publicKeyMultibase: z.string().optional(),
+})
+
+export const service = z.object({
+  id: z.string(),
+  type: z.string(),
+  serviceEndpoint: z.union([z.string(), z.record(z.unknown())]),
+})
+
+export const didDocument = z.object({
+  id: z.string(),
+  alsoKnownAs: z.array(z.string()).optional(),
+  verificationMethod: z.array(verificationMethod).optional(),
+  service: z.array(service).optional(),
+})
+
+export type DidDocument = z.infer<typeof didDocument>
+
+export function isValidDidDoc(doc: unknown): doc is DidDocument {
+  return didDocument.safeParse(doc).success
+}
+
+export function getPdsEndpoint(doc: unknown): URL | undefined {
+  if (isValidDidDoc(doc)) {
+    const pds = doc.service?.find((s) => s.type === 'AtprotoPersonalDataServer')
+    if (pds && typeof pds.serviceEndpoint === 'string') {
```

Comment:
Note: this is currently taking the first PDS service entry and only handling the string value. I don't know what a non-string value would look like.

Reply:
yeah that'd be considered a valid json-ld did document, but not supported by atproto rn

this is similar to how we validate services in our identity resolver 👍